virttest/utils_test/qemu.py

```diff
@@ -1517,100 +1517,125 @@ def action_after_suspend(self, **args):
         pass
 
 
-class MemoryBase(object):
+class MemoryBaseTest(object):
 
     """
     Base class for memory functions.
     """
 
+    UNIT = "MB"
+
     def __init__(self, test, params, env):
         self.test = test
         self.env = env
         self.params = params
         self.sessions = {}
 
+    def get_vm_mem(self, vm):
+        """
+        Count memory assigned to VM.
+
+        :param vm: VM object
+        :return: memory size in MB.
+        """
+        PC_DIMM = qdevices.Dimm
+        # Default memory size in configuration is MB, so append
+        # 'MB' in the end of 'mem' param.
+        mem_str = "%sM" % vm.params.get("mem", "0")
+        total_mem = self.normalize_mem_size(mem_str)
+        pc_dimms = filter(lambda x: isinstance(x, PC_DIMM), vm.devices)
+        obj_ids = map(lambda x: x.get_param('memdev'), pc_dimms)
+        obj_devs = map(lambda x: vm.devices.get_by_qid(x)[0], obj_ids)
+        obj_size = map(lambda x: x.get_param('size'), obj_devs)
+        total_mem += sum(map(self.normalize_mem_size, obj_size))
+        logging.info("Assigned %s%s " % (total_mem, self.UNIT) +
+                     "memory to '%s'" % vm.name)
+        return total_mem
+
     @classmethod
-    def get_guest_total_mem(cls, os_type, session):
+    def normalize_mem_size(cls, str_size):
         """
-        get physical memory size detect by guest os
+        Convert memory size unit
 
-        :param os_type: Guest os type.
-        :param session: VM session.
-        :return mem_size: total physical memory in guest.
+        :param str_size: memory size string, like: 1GB
+        :return: memory size value in MB
         """
-        error_context.context("Get physical memory in guest")
-        mem_size = 0.0
-        if os_type == "windows":
-            cmd = 'systeminfo | findstr /C:"Total Physical Memory"'
-            regex = re.compile("Total Physical Memory:\s+(.*)", re.M | re.I)
-        else:
-            cmd = 'cat /proc/meminfo'
-            regex = re.compile("memtotal:\s+(.*)", re.M | re.I)
-        output = session.cmd_output_safe(cmd, timeout=240)
+        args = (str_size, cls.UNIT, 1024)
         try:
-            mem_str = regex.search(output.replace(',', '')).groups()[0]
-            args = ("%sB" % mem_str, "M", 1024)
-            mem_size = float(utils_misc.normalize_data_size(*args))
-            logging.info("total memory reported by os: %.2f MB" % mem_size)
-        except Exception:
-            logging.warn("invalid output from guest: %s" % output)
-        return mem_size
+            size = utils_misc.normalize_data_size(*args)
+            return int(float(size))
+        except ValueError, details:
+            logging.debug("Convert memory size error('%s')" % details)
+        return 0
 
     @classmethod
-    def get_guest_free_mem(cls, os_type, session):
+    def get_guest_total_mem(cls, vm):
```

Comment:
get_vm_mem() is get memory from parameters (or CML), get_total_mem() is getting total mem from guest via wmic.